### PR TITLE
Achor stateroot tests

### LINKED ISSUE #74: Anchor contract interact unit tests
Anchor contract interacts unit tests.

### PR DESCRIPTION
**Pull request by 0xsarvesh**
Fixes #74

⚠️ Merge after #56   
Anchor state root contract interact.

### CODE REVIEW

File: src/ContractInteract/EIP20CoGateway.js [hunk 1]
```diff
@@ -316,6 +316,48 @@ class EIP20CoGateway {
...
+      return Promise.reject(err);
+    }
+
+    if (typeof unlockSecret !== 'string') {
+      const err = new TypeError('Invalid unlock secret.');
+      return Promise.reject(err);
+    }
+
+    const tx = this.contract.methods.progressRedeem(messageHash, unlockSecret);
+    return Promise.resolve(tx);
```
Comment (schemar) [inline src/ContractInteract/EIP20CoGateway.js]: You could add a comment `FIXME: https://github.com/OpenSTFoundation/mosaic.js/issues/64`

File: src/ContractInteract/EIP20Gateway.js [hunk 2]
```diff
@@ -535,8 +705,8 @@ class EIP20Gateway {
    */
   getLatestAnchorInfo() {
     return this.getAnchor().then((anchor) => {
-      anchor.getLatestStateRootBlockHeight().then((blockHeight) => {
-        anchor.getStateRoot(blockHeight).then((stateRoot) => {
+      return anchor.getLatestStateRootBlockHeight().then((blockHeight) => {
+        return anchor.getStateRoot(blockHeight).then((stateRoot) => {
```
Comment (schemar) [inline src/ContractInteract/EIP20Gateway.js]: See https://github.com/OpenSTFoundation/mosaic.js/pull/56/files#diff-2046288c7009bc2988ab960e2a5cf646R599 you probably need to rebase.

**Review by schemar — COMMENTED** [review]
Please rebase or merge develop

File: test/Anchor/anchorStateRoot.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// ----------------------------------------------------------------------------
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]: Please remove all license headers

File: test/Anchor/anchorStateRoot.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+const chai = require('chai');
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]:
```suggestion
const { assert } = require('chai');
```
?

File: test/Anchor/anchorStateRoot.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+const Web3 = require('web3');
+const sinon = require('sinon');
+
+const assert = chai.assert;
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]:
```suggestion
```
Delete if using deconstruction above.

File: test/Anchor/anchorStateRoot.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  it('should pass when called with correct arguments', async () => {
+    let blockHeight = '10';
+    let stateRoot = web3.utils.sha3('1');
+    let mockTx = 'mockTx';
+
+    let txOptions = {
+      from: '0x0000000000000000000000000000000000000003',
+    };
+
+    let spyMethod = sinon.replace(
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]: maybe?
```suggestion
    let spyAnchorStateRoot = sinon.replace(
```

File: test/Anchor/anchorStateRoot.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      from: '0x0000000000000000000000000000000000000003',
+    };
+
+    await AssertAsync.reject(
+      anchor.anchorStateRoot(blockHeight, stateRoot, txOptions),
+      `Invalid block height: ${blockHeight}.`,
+    );
+  });
+
+  it('should throw for undefined block height', async () => {
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]:
```suggestion
  it('should throw for undefined state root', async () => {
```

File: test/Anchor/anchorStateRoot.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+    let txOptions = undefined;
+
+    await AssertAsync.reject(
+      anchor.anchorStateRoot(blockHeight, stateRoot, txOptions),
+      `Invalid transaction options: ${txOptions}.`,
+    );
+  });
+
+  it('should throw for from address in txOptions is invalid', async () => {
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]:
```suggestion
  it('should throw if "from" address in txOptions is invalid', async () => {
```

File: test/Anchor/anchorStateRoot.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    let txOptions = {
+      from: '0x123',
+    };
+    await AssertAsync.reject(
+      anchor.anchorStateRoot(blockHeight, stateRoot, txOptions),
+      `Invalid from address: ${txOptions.from}.`,
+    );
+  });
+
+  it('should throw for from address in txOptions is undefined', async () => {
```
Comment (schemar) [inline test/Anchor/anchorStateRoot.js]:
```suggestion
  it('should throw if "from" address in txOptions is undefined', async () => {
```

File: test/Anchor/constructor.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const chai = require('chai');
+const Web3 = require('web3');
+const sinon = require('sinon');
+
+const assert = chai.assert;
```
Comment (schemar) [inline test/Anchor/constructor.js]: same as above?

File: test/Anchor/constructor.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+  beforeEach(() => {
+    web3 = new Web3();
+    anchorAddress = '0x0000000000000000000000000000000000000002';
+  });
+
+  it('should pass when called with correct arguments', async () => {
+    let mockObject = sinon.fake();
+
+    spyContract = sinon.replace(
```
Comment (schemar) [inline test/Anchor/constructor.js]: Maybe? Better readable.
```suggestion
    spyAnchor = sinon.replace(
```

File: test/Anchor/constructor.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+      Contracts,
+      'getAnchor',
+      sinon.fake.returns(mockObject),
+    );
+
+    const contractObject = new Anchor(web3, anchorAddress);
```
Comment (schemar) [inline test/Anchor/constructor.js]:
```suggestion
    const anchor = new Anchor(web3, anchorAddress);
```

File: test/Anchor/constructor.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    assert.strictEqual(
+      contractObject.anchorAddress,
```
Comment (schemar) [inline test/Anchor/constructor.js]: Wasn't this renamed to `anchor.address` in another PR? 🤔

File: test/Anchor/constructor.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+
+    SpyAssert.assert(spyContract, 1, [[web3, anchorAddress]]);
+
+    sinon.restore();
+  });
+
+  it('should throw an error when web3 object is undefined', async () => {
+    assert.throws(() => {
+      new Anchor(undefined, anchorAddress);
+    }, /Mandatory Parameter 'web3' is missing or invalid/);
```
Comment (schemar) [inline test/Anchor/constructor.js:48]: Weird line breaks. On purpose? Also below.

File: test/Anchor/constructor.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  });
+
+  it('should throw an error when anchor contract address is undefined', async () => {
+    assert.throws(() => {
+      new Anchor(web3, undefined);
+    }, /Mandatory Parameter 'anchorAddress' is missing or invalid./);
+  });
+
+  it('should throw error when contract instance is undefined', function() {
+    spyContract = sinon.replace(
```
Comment (schemar) [inline test/Anchor/constructor.js]:
```suggestion
    spyAnchor = sinon.replace(
```

File: test/Anchor/getLatestStateRootBlockHeight.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const chai = require('chai');
+const Web3 = require('web3');
+const sinon = require('sinon');
+
+const assert = chai.assert;
```
Comment (schemar) [inline test/Anchor/getLatestStateRootBlockHeight.js]: same

File: test/Anchor/getStateRoot.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+//
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+const chai = require('chai');
+const Web3 = require('web3');
+const sinon = require('sinon');
+
+const assert = chai.assert;
```
Comment (schemar) [inline test/Anchor/getStateRoot.js]: ☝️

**Review by schemar — CHANGES_REQUESTED** [review]
Nice!! 🐱 

Structure looks super good.
Only styling/typo comments.

File: test/Anchor/constructor.js (hunk 13 above)
Comment (hobofan) [reply test/Anchor/constructor.js]: Yes, in #58, but that's not merged yet.

**Comment by schemar** [comment]
Please add `use strict`; to the beginning of all files. Sorry I forgot 👶

File: test/Anchor/constructor.js (hunk 14 above)
Comment (schemar) [reply test/Anchor/constructor.js:48]: I guess yes 😅

**Review by schemar — APPROVED** [review]
:shipit:

**Review by deepesh-kn — APPROVED** [review]
LGTM 👍

**Pull request merged**